File: packages/azure-servicebus/src/producer.ts
/**
 * Azure Service Bus Producer Implementation
 */

import { ServiceBusClient, type ServiceBusSender, type ServiceBusMessage } from '@azure/service-bus';
import {
  BaseProducer,
  type PublishOptions,
  type HealthStatus,
  ConnectionError,
  PublishError,
  generateUUID,
} from '@anyq/core';
import type { ServiceBusProducerConfig } from './config.js';

/**
 * Azure Service Bus Producer
 *
 * Publishes messages to Azure Service Bus queues or topics
 */
export class ServiceBusProducer<T = unknown> extends BaseProducer<T> {
  private client: ServiceBusClient | null = null;
  private sender: ServiceBusSender | null = null;
  private readonly serviceBusConfig: ServiceBusProducerConfig;

  constructor(config: ServiceBusProducerConfig) {
    super(config);
    this.serviceBusConfig = config;

    // Validate config - must have either queue or topic
    if (!config.queue && !config.topic) {
      throw new Error('ServiceBusProducerConfig must specify either queue or topic');
    }
  }

  /**
   * Connect to Azure Service Bus
   */
  async connect(): Promise<void> {
    if (this._connected) {
      return;
    }

    try {
      const { connection, queue, topic } = this.serviceBusConfig;

      this.client = new ServiceBusClient(connection.connectionString);

      // Create sender for queue or topic
      const entityPath = topic?.name ?? queue?.name;
      if (!entityPath) {
        throw new Error('No queue or topic name specified');
      }

      this.sender = this.client.createSender(entityPath);

      this._connected = true;
      this.logger.info('Azure Service Bus producer connected', {
        entity: entityPath,
        type: topic ? 'topic' : 'queue',
      });
    } catch (error) {
      throw new ConnectionError(
        'Failed to connect to Azure Service Bus',
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Disconnect from Azure Service Bus
   */
  async disconnect(): Promise<void> {
    try {
      if (this.sender) {
        await this.sender.close();
        this.sender = null;
      }

      if (this.client) {
        await this.client.close();
        this.client = null;
      }

      this._connected = false;
      this.logger.info('Azure Service Bus producer disconnected');
    } catch (error) {
      this.logger.error('Error during disconnect', { error });
    }
  }

  /**
   * Publish a single message
   */
  async publish(body: T, options?: PublishOptions): Promise<string> {
    if (!this.sender) {
      throw new ConnectionError('Producer not connected');
    }

    const messageId = generateUUID();

    try {
      const serializedBody = this.serializer.serialize(body);

      const message: ServiceBusMessage = {
        messageId,
        body: serializedBody,
        contentType: 'application/json',
        sessionId: options?.groupId, // Map groupId to sessionId
        partitionKey: options?.key, // Map key to partitionKey
        correlationId: options?.correlationId,
        replyTo: options?.replyTo,
        timeToLive: options?.ttlMs,
        applicationProperties: options?.headers as Record<string, string | number | boolean | Date>,
      };

      // Add scheduled enqueue time if delay is specified
      if (options?.delaySeconds) {
        message.scheduledEnqueueTimeUtc = new Date(Date.now() + options.delaySeconds * 1000);
      }

      await this.sender.sendMessages(message);

      this.logger.debug('Message published', {
        messageId,
        sessionId: options?.groupId,
        partitionKey: options?.key,
      });

      return messageId;
    } catch (error) {
      throw new PublishError('Failed to publish message', {
        cause: error instanceof Error ? error : undefined,
      });
    }
  }

  /**
   * Publish multiple messages
   */
  async publishBatch(
    messages: Array<{ body: T; options?: PublishOptions }>
  ): Promise<string[]> {
    if (!this.sender) {
      throw new ConnectionError('Producer not connected');
    }

    const messageIds: string[] = [];

    try {
      // Create a batch
      const batch = await this.sender.createMessageBatch();

      for (const { body, options } of messages) {
        const messageId = generateUUID();
        messageIds.push(messageId);

        const serializedBody = this.serializer.serialize(body);

        const message: ServiceBusMessage = {
          messageId,
          body: serializedBody,
          contentType: 'application/json',
          sessionId: options?.groupId,
          partitionKey: options?.key,
          correlationId: options?.correlationId,
          replyTo: options?.replyTo,
          timeToLive: options?.ttlMs,
          applicationProperties: options?.headers as Record<string, string | number | boolean | Date>,
        };

        if (options?.delaySeconds) {
          message.scheduledEnqueueTimeUtc = new Date(Date.now() + options.delaySeconds * 1000);
        }

        const added = batch.tryAddMessage(message);
        if (!added) {
          // Batch is full, send it and create a new one
          await this.sender.sendMessages(batch);
          const newBatch = await this.sender.createMessageBatch();
          newBatch.tryAddMessage(message);
        }
      }

      // Send remaining messages
      if (batch.count > 0) {
        await this.sender.sendMessages(batch);
      }

      this.logger.debug('Batch published', {
        count: messageIds.length,
      });

      return messageIds;
    } catch (error) {
      throw new PublishError('Failed to publish batch', {
        cause: error instanceof Error ? error : undefined,
      });
    }
  }

  /**
   * Flush pending messages
   */
  async flush(): Promise<void> {
    // Azure Service Bus sends immediately, no flush needed
  }

  /**
   * Get health status
   */
  async healthCheck(): Promise<HealthStatus> {
    const start = Date.now();

    try {
      if (!this._connected || !this.sender) {
        return {
          healthy: false,
          connected: false,
          error: 'Not connected',
        };
      }

      // Check if sender is open
      const entityPath = this.serviceBusConfig.topic?.name ?? this.serviceBusConfig.queue?.name;

      return {
        healthy: true,
        connected: true,
        latencyMs: Date.now() - start,
        details: {
          entity: entityPath,
          type: this.serviceBusConfig.topic ? 'topic' : 'queue',
        },
      };
    } catch (error) {
      return {
        healthy: false,
        connected: false,
        latencyMs: Date.now() - start,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

/**
 * Create an Azure Service Bus producer instance
 */
export function createServiceBusProducer<T = unknown>(
  config: ServiceBusProducerConfig
): ServiceBusProducer<T> {
  return new ServiceBusProducer<T>(config);
}

File: go/azureservicebus/producer.go
package azureservicebus

import (
	"context"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"

	"github.com/sns45/anyq/go/core"
)

// Producer publishes messages to an Azure Service Bus queue or topic.
type Producer struct {
	core.BaseProducer
	cfg    Config
	client *azservicebus.Client
	sender *azservicebus.Sender
	entity string
}

// NewProducer builds an Azure Service Bus producer.
func NewProducer(cfg Config) *Producer {
	if cfg.Driver == "" {
		cfg.Driver = core.DriverAzureServiceBus
	}
	p := &Producer{cfg: cfg, entity: cfg.senderEntity()}
	p.InitBase(cfg.BaseQueueConfig)
	return p
}

// Connect creates the Service Bus client and a sender for the queue/topic.
func (p *Producer) Connect(ctx context.Context) error {
	if p.IsConnected() {
		return nil
	}
	if p.cfg.Queue == nil && p.cfg.Topic == nil {
		return core.NewConnectionError("azure service bus producer requires a queue or topic", nil)
	}
	if p.cfg.Connection.ConnectionString == "" {
		return core.NewConnectionError("azure service bus producer requires a connection string", nil)
	}

	client, err := azservicebus.NewClientFromConnectionString(p.cfg.Connection.ConnectionString, nil)
	if err != nil {
		return core.NewConnectionError("failed to connect to Azure Service Bus", err)
	}

	entity := p.cfg.senderEntity()
	if entity == "" {
		_ = client.Close(ctx)
		return core.NewConnectionError("no queue or topic name specified", nil)
	}

	sender, err := client.NewSender(entity, nil)
	if err != nil {
		_ = client.Close(ctx)
		return core.NewConnectionError("failed to create Service Bus sender", err)
	}

	p.client = client
	p.sender = sender
	p.entity = entity
	p.SetConnected(true)
	kind := "queue"
	if p.cfg.Topic != nil {
		kind = "topic"
	}
	p.Logger.Info("Azure Service Bus producer connected", map[string]any{"entity": entity, "type": kind})
	return nil
}

// Disconnect closes the sender and client.
func (p *Producer) Disconnect(ctx context.Context) error {
	if !p.IsConnected() {
		return nil
	}
	if p.sender != nil {
		if err := p.sender.Close(ctx); err != nil {
			p.Logger.Error("error closing sender", map[string]any{"error": err.Error()})
		}
		p.sender = nil
	}
	if p.client != nil {
		if err := p.client.Close(ctx); err != nil {
			p.Logger.Error("error closing client", map[string]any{"error": err.Error()})
		}
		p.client = nil
	}
	p.SetConnected(false)
	p.Logger.Info("Azure Service Bus producer disconnected", nil)
	return nil
}

// Publish sends a single message and returns its message id.
func (p *Producer) Publish(ctx context.Context, body []byte, opts *core.PublishOptions) (string, error) {
	if p.sender == nil {
		return "", core.NewConnectionError("producer not connected", nil)
	}
	msg, id := buildMessage(body, opts)
	if err := p.sender.SendMessage(ctx, msg, nil); err != nil {
		return "", core.NewPublishError("failed to publish message", true, err)
	}
	p.Logger.Debug("message published", map[string]any{"messageId": id, "entity": p.entity})
	return id, nil
}

// PublishBatch sends multiple messages, packing them into SDK message batches.
func (p *Producer) PublishBatch(ctx context.Context, items []core.BatchItem) ([]string, error) {
	if p.sender == nil {
		return nil, core.NewConnectionError("producer not connected", nil)
	}

	ids := make([]string, 0, len(items))
	batch, err := p.sender.NewMessageBatch(ctx, nil)
	if err != nil {
		return nil, core.NewPublishError("failed to create message batch", true, err)
	}

	for _, item := range items {
		msg, id := buildMessage(item.Body, item.Options)
		addErr := batch.AddMessage(msg, nil)
		if addErr == azservicebus.ErrMessageTooLarge {
			// Current batch is full: send it, start a new one, and re-add.
			if batch.NumMessages() == 0 {
				return ids, core.NewPublishError("message too large for an empty batch", false, addErr)
			}
			if err := p.sender.SendMessageBatch(ctx, batch, nil); err != nil {
				return ids, core.NewPublishError("failed to publish batch", true, err)
			}
			batch, err = p.sender.NewMessageBatch(ctx, nil)
			if err != nil {
				return ids, core.NewPublishError("failed to create message batch", true, err)
			}
			if addErr = batch.AddMessage(msg, nil); addErr != nil {
				return ids, core.NewPublishError("failed to add message to batch", false, addErr)
			}
		} else if addErr != nil {
			return ids, core.NewPublishError("failed to add message to batch", false, addErr)
		}
		ids = append(ids, id)
	}

	if batch.NumMessages() > 0 {
		if err := p.sender.SendMessageBatch(ctx, batch, nil); err != nil {
			return ids, core.NewPublishError("failed to publish batch", true, err)
		}
	}

	p.Logger.Debug("batch published", map[string]any{"count": len(ids), "entity": p.entity})
	return ids, nil
}

// Flush is a no-op: Azure Service Bus sends messages immediately.
func (p *Producer) Flush(ctx context.Context) error { return nil }

// HealthCheck reports producer health.
func (p *Producer) HealthCheck(ctx context.Context) (core.HealthStatus, error) {
	start := time.Now()
	if p.sender == nil || !p.IsConnected() {
		return core.HealthStatus{Healthy: false, Connected: false, Error: "Not connected"}, nil
	}
	kind := "queue"
	if p.cfg.Topic != nil {
		kind = "topic"
	}
	return core.HealthStatus{
		Healthy:   true,
		Connected: true,
		LatencyMs: time.Since(start).Milliseconds(),
		Details:   map[string]any{"entity": p.entity, "type": kind},
	}, nil
}

// Client returns the underlying client (for testing).
func (p *Producer) Client() *azservicebus.Client { return p.client }

// buildMessage maps a body + PublishOptions to an azservicebus.Message, returning
// the generated message id.
func buildMessage(body []byte, opts *core.PublishOptions) (*azservicebus.Message, string) {
	id := core.GenerateMessageID()
	contentType := "application/json"
	msg := &azservicebus.Message{
		MessageID:   &id,
		Body:        body,
		ContentType: &contentType,
	}
	if opts == nil {
		return msg, id
	}

	// Map groupId to SessionID, key to PartitionKey (mirrors the TS adapter).
	if opts.GroupID != "" {
		gid := opts.GroupID
		msg.SessionID = &gid
	}
	if opts.Key != "" {
		pk := opts.Key
		msg.PartitionKey = &pk
	}
	if opts.CorrelationID != "" {
		cid := opts.CorrelationID
		msg.CorrelationID = &cid
	}
	if opts.ReplyTo != "" {
		rt := opts.ReplyTo
		msg.ReplyTo = &rt
	}
	if opts.TTL > 0 {
		ttl := opts.TTL
		msg.TimeToLive = &ttl
	}
	if len(opts.Headers) > 0 {
		props := make(map[string]any, len(opts.Headers))
		for k, v := range opts.Headers {
			props[k] = string(v)
		}
		msg.ApplicationProperties = props
	}
	// Native scheduled delivery for a publish-time delay.
	if opts.Delay > 0 {
		t := time.Now().Add(opts.Delay)
		msg.ScheduledEnqueueTime = &t
	}
	return msg, id
}

var _ core.Producer = (*Producer)(nil)
